DirectProgramming/DPC++/GraphTraversal/odd-even-merge-sort/sycl_migrated/src/odd_even_merge_sort.cpp

```diff
@@ -0,0 +1,221 @@
+
+  item.barrier();
+  d_DstKey[0] = s_key[item.get_local_id(2) + 0];
+  d_DstVal[0] = s_val[item.get_local_id(2) + 0];
+  d_DstKey[(SHARED_SIZE_LIMIT / 2)] =
+      s_key[item.get_local_id(2) + (SHARED_SIZE_LIMIT / 2)];
+  d_DstVal[(SHARED_SIZE_LIMIT / 2)] =
+      s_val[item.get_local_id(2) + (SHARED_SIZE_LIMIT / 2)];
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Odd-even merge sort iteration kernel
+// for large arrays (not fitting into shared memory)
+////////////////////////////////////////////////////////////////////////////////
+void oddEvenMergeGlobal(uint *d_DstKey, uint *d_DstVal, uint *d_SrcKey,
+                        uint *d_SrcVal, uint arrayLength, uint size,
+                        uint stride, uint dir, nd_item<3> item) {
+  uint global_comparatorI =
+      item.get_group(2) * item.get_local_range().get(2) + item.get_local_id(2);
+
+  // Odd-even merge
+  uint pos = 2 * global_comparatorI - (global_comparatorI & (stride - 1));
+
+  if (stride < size / 2) {
+    uint offset = global_comparatorI & ((size / 2) - 1);
+
+    if (offset >= stride) {
+      uint keyA = d_SrcKey[pos - stride];
+      uint valA = d_SrcVal[pos - stride];
+      uint keyB = d_SrcKey[pos + 0];
+      uint valB = d_SrcVal[pos + 0];
+
+      Comparator(keyA, valA, keyB, valB, dir);
+
+      d_DstKey[pos - stride] = keyA;
+      d_DstVal[pos - stride] = valA;
+      d_DstKey[pos + 0] = keyB;
+      d_DstVal[pos + 0] = valB;
+    }
+  } else {
+    uint keyA = d_SrcKey[pos + 0];
+    uint valA = d_SrcVal[pos + 0];
+    uint keyB = d_SrcKey[pos + stride];
+    uint valB = d_SrcVal[pos + stride];
+
+    Comparator(keyA, valA, keyB, valB, dir);
+
+    d_DstKey[pos + 0] = keyA;
+    d_DstVal[pos + 0] = valA;
+    d_DstKey[pos + stride] = keyB;
+    d_DstVal[pos + stride] = valB;
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Interface function
+////////////////////////////////////////////////////////////////////////////////
+// Helper function
+uint factorRadix2(uint *log2L, uint L) {
+  if (!L) {
+    *log2L = 0;
+    return 0;
+  } else {
+    for (*log2L = 0; (L & 1) == 0; L >>= 1, log2L++)
+      ;
+    return L;
+  }
+}
+
+extern "C" uint oddEvenMergeSort(uint *d_DstKey, uint *d_DstVal, uint *d_SrcKey,
+                                 uint *d_SrcVal, uint batchSize,
+                                 uint arrayLength, uint dir, queue &q) {
+  // Nothing to sort
+  if (arrayLength < 2) return 0;
+
+  // Only power-of-two array lengths are supported by this implementation
+  uint log2L;
+  uint factorizationRemainder = factorRadix2(&log2L, arrayLength);
+  assert(factorizationRemainder == 1);
+
+  dir = (dir != 0);
+
+  uint num_workgroups = (batchSize * arrayLength) / SHARED_SIZE_LIMIT;
+  uint workgroup_size = SHARED_SIZE_LIMIT / 2;
+
+  if (arrayLength <= SHARED_SIZE_LIMIT) {
+    assert(SHARED_SIZE_LIMIT % arrayLength == 0);
+
+    q.submit([&](handler &cgh) {
```

Comment:
Since this is manually migrated you could use "h" here instead of "cgh".

Reply:
Updated as per the feedback.